### DISCUSSION TITLE
Python cannot change values in an internal/nested struct

### DISCUSSION THREAD

**Discussion by sdfgeoff** [PyO3/pyo3#1974] [Questions] [answered]
### Bug Description

Using the `#[pyo3(get, set)]` only works if python directly owns the class, but not if the struct is contained inside another rust/py03 object.

The same issue is present if using a function (in the inner class) to get/set the variable. I suspect that python is cloning the inner object and throwing it away.

### Steps to Reproduce

Consider the Rust structs:
```rust
#[pyclass]
struct PainterCore {
    #[pyo3(get, set)]
    pub test1: f32,
    #[pyo3(get, set)]
    pub inner: InnerStruct,
}

#[pymethods]
impl PainterCore {
    #[new]
    pub fn new() -> PyResult<Self> {
        Ok(Self {
            test1: 0.0,
            inner: InnerStruct{
                test2: 0.0
            }
        })
    }
}

#[pyclass]
#[derive(Clone)]
struct InnerStruct {
    #[pyo3(get, set)]
    test2: f32,
}
```

And consider the python:
```python
        self.core = painter_core.PainterCore()

        print(self.core.test1, self.core.inner.test2)
        self.core.test1 = 5.0
        self.core.inner.test2 = 5.0
        print(self.core.test1, self.core.inner.test2)
```

The expected output from the first print is: `0.0 0.0` because variables `test1` and `test2` have been initialised to zero. The expected output from the second print is `5.0 5.0` because both variables have been set to `5.0`

The actual output of the second print is `5.0 0.0`

### Your operating system and version

Archlinux

### Your Python version (`python --version`)

3.9.7

### Your Rust version (`rustc --version`)

rustc 1.55.0 (c8dfcfe04 2021-09-06)

### Your PyO3 version

0.14.5

### How did you install python? Did you use a virtualenv?
```bash
sudo pacman -S python
```
```make
# Makefile
VENV_START = ${PYTHON} -m venv .env; source .env/bin/activate;

setup:
	${VENV_START} \
	pip install -r requirements.txt; \
```

```
# requirements.txt
maturin
```

**Comment by birkenfeld** [comment, marked as answer]
The attribute access `self.core.inner` returns a copy of the object; any changes you make will not be reflected there.

If you need to return a reference instead, the struct member has to be `Py<InnerStruct>`.

**Comment by mejrs** [comment]
See also https://pyo3.rs/v0.15.0/faq.html#pyo3get-clones-my-field

**Reply by Trolldemorted** [reply]
I have just run into this and it was quite unexpected. Shouldn't this be mentioned in the [documentation to classes](https://pyo3.rs/v0.18.3/class)?

This is kind of a footgun, and neither the class documentation nor the conversion table mention it.
